### PR TITLE
Fix for the hang during deletion of engine=Kafka (one more time)

### LINKED ISSUE #7260: Kafka: drop Kafka table sometimes hangs
stacktrace at the moment of hang.
https://gist.github.com/filimonov/c51e77f50b4eebb43a133c6cfc4d9066

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix for the hang which was happening sometimes during DROP of table engine=Kafka (or during server restarts).

Detailed description / Documentation draft:
Continuation of https://github.com/ClickHouse/ClickHouse/pull/10910
After #10910 it still can hang, but chances get lower and the test stopped detecting that.
librdkafka have a flag that should prevent that, so I've added it to cppkafka (https://github.com/mfontanini/cppkafka/pull/247) 
But after extra testing, I've found that there are scenarios when that flag don't work (https://github.com/edenhill/librdkafka/issues/2898) and we still need to get back to queue draining after unsubscribe. 

Fixes #7260 #10740

Thanks to @azat for help & inspiration.

### CODE REVIEW

File: tests/integration/test_storage_kafka/test.py [hunk 1]
```diff
@@ -246,6 +246,41 @@ def test_kafka_consumer_hang(kafka_cluster):
...
+                     kafka_group_name = 'consumer_hang2',
+                     kafka_format = 'JSONEachRow';
+        ''')
+
+    instance.query('SELECT * FROM test.kafka')
+    instance.query('SELECT * FROM test.kafka2')
+#echo 'SELECT * FROM test.kafka; SELECT * FROM test.kafka2; DROP TABLE test.kafka;' | clickhouse client -mn &
+#    kafka_cluster.open_bash_shell('instance')
+
+    instance.query('DROP TABLE test.kafka')
```
Comment (azat) [inline tests/integration/test_storage_kafka/test.py:275]: maybe left some comments on why this reproduces the issue?

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 2]
```diff
@@ -80,9 +81,65 @@ ReadBufferFromKafkaConsumer::ReadBufferFromKafkaConsumer(
...
+            else
+            {
+                LOG_WARNING(log, "Error during ReadBufferFromKafkaConsumer draining: " << err);
+            }
+        }
+
+        auto ts = std::chrono::steady_clock::now();
+        if (std::chrono::duration_cast<std::chrono::milliseconds>(ts-start_time) > DRAIN_TIMEOUT_MS)
+        {
+            break;
```
Comment (azat) [inline src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:138]: definitely needs a message

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 3]
```diff
@@ -80,9 +81,66 @@ ReadBufferFromKafkaConsumer::ReadBufferFromKafkaConsumer(
...
+
+    while (true)
+    {
+        auto msg = consumer->poll(100ms);
+        if (!msg)
+            break;
+
+        if (auto err = msg.get_error())
+        {
+            if (msg.is_eof() || err == last_error)
```
[the remark this answers is not in the document]
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:125]: Thanks for the catch. I loose one line.

[the remark this answers is not in the document]
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp:125]: RD_KAFKA_RESP_ERR_NO_ERROR is catched by if. I was really wanted to check if we get the same error more than once.

File: src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp [hunk 4]
```diff
@@ -80,9 +81,66 @@ ReadBufferFromKafkaConsumer::ReadBufferFromKafkaConsumer(
...
+            catch (const cppkafka::HandleException & e)
+            {
+                LOG_ERROR(log, "Exception from ReadBufferFromKafkaConsumer destructor (unsubscribe): " << e.what());
+            }
+            drain();
+        }
+    }
+    catch (const cppkafka::HandleException & e)
+    {
+        LOG_ERROR(log, "Exception from ReadBufferFromKafkaConsumer destructor: " << e.what());
```
[the remark this answers is not in the document]
Comment (filimonov) [reply src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp]: check now

**Review by azat — APPROVED** [review]
LGTM (squashing before merging is preferable)

But

>Fix for the hang during deletion of engine=Kafka (one more time) #11145

"one more time" looks icky, let's change PR title (and the commit subject) so that it will include "RD_KAFKA_DESTROY_F_NO_CONSUMER_CLOSE" and "get back poll" or something similar

Besides commit subject is too long (I guess you missed newline between subject and description)

**Pull request merged**